packages/components/src/tooltip/test/index.js

```diff
@@ -152,6 +152,40 @@ describe( 'Tooltip', () => {
 			}, TOOLTIP_DELAY );
 		} );
 
+		it( 'should respect custom delay prop when showing popover', () => {
+			const originalMouseEnter = jest.fn();
+			jest.useFakeTimers();
+			const wrapper = mount(
+				<Tooltip text="Help text" delay={ 2000 }>
+					<button
+						onMouseEnter={ originalMouseEnter }
+						onFocus={ originalMouseEnter }
+					>
+						<span>Hover Me!</span>
+					</button>
+				</Tooltip>
+			);
+
+			const button = wrapper.find( 'button' );
+			button.simulate( 'mouseenter', { type: 'mouseenter' } );
+
+			const popoverBeforeTimeout = wrapper.find( 'Popover' );
+			expect( popoverBeforeTimeout ).toHaveLength( 0 );
+			expect( originalMouseEnter ).toHaveBeenCalledTimes( 1 );
+
+			// Popover does not yet exist after default delay, because custom delay is passed.
+			setTimeout( () => {
+				const popoverBetweenTimeout = wrapper.find( 'Popover' );
+				expect( popoverBetweenTimeout ).toHaveLength( 0 );
+			}, TOOLTIP_DELAY );
+
+			// Popover appears after custom delay.
+			setTimeout( () => {
+				const popoverAfterTimeout = wrapper.find( 'Popover' );
+				expect( popoverAfterTimeout ).toHaveLength( 1 );
+			}, 2000 );
+		} );
```

Comment:
Do we need to restore timers again after the test? Re-checking the docs, we should probably do this in an `afterEach()` as it seems `jest.useFakeTimers();` is [a global operation](https://jestjs.io/docs/timer-mocks).

```
jest.runOnlyPendingTimers()
jest.useRealTimers()
```

Reply:
@getdave I did not fully understand why `jest.runOnlyPendingTimers()` is needed. Just `jest.useRealTimers()` enough in `afterEach()`

Reply:
@getdave I tried your suggestion in local. Below is the diff I tried on top of this PR. But could not get the tests to pass.

#### Diff:
```diff
diff --git a/packages/components/src/tooltip/test/index.js b/packages/components/src/tooltip/test/index.js
index d3648904ef..163563e317 100644
--- a/packages/components/src/tooltip/test/index.js
+++ b/packages/components/src/tooltip/test/index.js
@@ -15,6 +15,10 @@ import { act } from '@testing-library/react';
 
 describe( 'Tooltip', () => {
 	describe( '#render()', () => {
+		afterEach( () => {
+			jest.useRealTimers();
+		} );
+
 		it( 'should render children (abort) if multiple children passed', () => {
 			// Mount: Enzyme shallow does not support wrapping multiple nodes
 			const wrapper = mount(
@@ -146,10 +150,9 @@ describe( 'Tooltip', () => {
 			expect( originalMouseEnter ).toHaveBeenCalledTimes( 1 );
 
 			// Force delayedSetIsOver to be called
-			setTimeout( () => {
-				const popoverAfterTimeout = wrapper.find( 'Popover' );
-				expect( popoverAfterTimeout ).toHaveLength( 1 );
-			}, TOOLTIP_DELAY );
+			jest.advanceTimersByTime( TOOLTIP_DELAY );
+			const popoverAfterTimeout = wrapper.find( 'Popover' );
+			expect( popoverAfterTimeout ).toHaveLength( 1 );
 		} );
 
 		it( 'should respect custom delay prop when showing popover', () => {
@@ -174,19 +177,19 @@ describe( 'Tooltip', () => {
 			expect( originalMouseEnter ).toHaveBeenCalledTimes( 1 );
 
 			// Popover does not yet exist after default delay, because custom delay is passed.
-			setTimeout( () => {
-				const popoverBetweenTimeout = wrapper.find( 'Popover' );
-				expect( popoverBetweenTimeout ).toHaveLength( 0 );
-			}, TOOLTIP_DELAY );
+			jest.advanceTimersByTime( TOOLTIP_DELAY );
+			const popoverBetweenTimeout = wrapper.find( 'Popover' );
+			expect( popoverBetweenTimeout ).toHaveLength( 0 );
 
 			// Popover appears after custom delay.
-			setTimeout( () => {
-				const popoverAfterTimeout = wrapper.find( 'Popover' );
-				expect( popoverAfterTimeout ).toHaveLength( 1 );
-			}, 2000 );
+			jest.advanceTimersByTime( 2000 - TOOLTIP_DELAY );
+			const popoverAfterTimeout = wrapper.find( 'Popover' );
+			expect( popoverAfterTimeout ).toHaveLength( 1 );
 		} );
 
 		it( 'should show tooltip when an element is disabled', () => {
+			jest.useFakeTimers();
+
 			const wrapper = mount(
 				<Tooltip text="Show helpful text here">
 					<button disabled>Click me</button>
@@ -205,10 +208,9 @@ describe( 'Tooltip', () => {
 				currentTarget: {},
 			} );
 
-			setTimeout( () => {
-				const popover = wrapper.find( 'Popover' );
-				expect( popover ).toHaveLength( 1 );
-			}, TOOLTIP_DELAY );
+			jest.advanceTimersByTime( TOOLTIP_DELAY );
+			const popover = wrapper.find( 'Popover' );
+			expect( popover ).toHaveLength( 1 );
 		} );
 
 		it( 'should cancel pending setIsOver on mouseleave', () => {
```

I am getting below error, which suggests that the timers are not elapsing.

```
Expected length: 1
    Received length: 0
    Received object: {}

      211 |                     jest.advanceTimersByTime( TOOLTIP_DELAY );
      212 |                     const popover = wrapper.find( 'Popover' );
    > 213 |                     expect( popover ).toHaveLength( 1 );
          |                                       ^
      214 |             } );
      215 | 
      216 |             it( 'should cancel pending setIsOver on mouseleave', () => {
```

Can you tell what I am doing wrong here?

Reply:
> While you can call jest.useFakeTimers() or jest.useRealTimers() from anywhere (top level, inside an it block, etc.), it is a global operation and will affect other tests within the same file. 

The docs suggest it's a global operation and so it needs to be cleared in an after each for all the tests in your file even if you only call `useFakeTimers` in an individual `it()`.

Reply:
OK having read [this thread](https://bleepcoder.com/jest/226312984/calling-runalltimers-after-using-lodash-s-debounce-results) I think we can say that Jest doesn't handle mocking `setTimeout`'s too well.

It looks like many folks have been able to get good results using the method you've implemented so let's go with that for now.

However, if you get inconsistent results running these tests on CI then we might need to look at a more advanced timer mock library such as Sinon.

Reply:
> OK having read this thread I think we can say that Jest doesn't handle mocking lodash debounce too well because it utilises Date.now() and other methods alongside setTimeout for greater accuracy.

Yes! I observed below error (mentioned in the article you linked) and I thought, it's something I messed up, so did not follow that trail. 🤦🏻‍♂️ 

```
Ran 100000 timers, and there are still more! Assuming we've hit an infinite recursion and bailing out...
```